packages/browser-sdk/src/feature/features.ts

```diff
@@ -177,40 +178,60 @@ export class FeaturesClient {
   }
 
   private async maybeFetchFeatures(): Promise<APIFeaturesResponse | undefined> {
-    const cachedItem = this.cache.get(this.fetchParams().toString());
-
-    // if there's no cached item OR the cached item is a failure and we haven't retried
-    // too many times yet - fetch now
-    if (
-      !cachedItem ||
-      (!cachedItem.success &&
-        (this.config.failureRetryAttempts === false ||
-          cachedItem.attemptCount < this.config.failureRetryAttempts))
-    ) {
-      return await this.fetchFeatures();
-    }
+    const cacheKey = this.fetchParams().toString();
+    const cachedItem = this.cache.get(cacheKey);
+
+    if (cachedItem) {
+      if (!cachedItem.stale) return cachedItem.features;
 
-    // cachedItem is a success or a failed attempt that we've retried too many times
-    if (cachedItem.stale) {
       // serve successful stale cache if `staleWhileRevalidate` is enabled
-      if (this.config.staleWhileRevalidate && cachedItem.success) {
-        // re-fetch in the background, return last successful value
-        this.fetchFeatures().catch(() => {
-          // we don't care about the result, we just want to re-fetch
-        });
+      if (this.config.staleWhileRevalidate) {
+        // re-fetch in the background, but immediately return last successful value
+        this.fetchFeatures()
+          .then((features) => {
+            if (!features) return;
+
+            this.cache.set(cacheKey, {
+              features,
+            });
+            this.setFeatures(features);
+          })
+          .catch(() => {
+            // we don't care about the result, we just want to re-fetch
+          });
         return cachedItem.features;
       }
+    }
+
+    // if there's no cached item or there is a stale one but `staleWhileRevalidate` is disabled
+    // try fetching a new one
+    const fetchedFlags = await this.fetchFeatures();
```

Comment:
I see a mixture of "flags" and "features" in the names. Probably better to switch to features in every name and comment for clarity

Reply:
good catch. Missed that one 👍